### PR TITLE
[expo-updates][docs] fix development mode crash on iOS, add documentation for testing

### LINKED ISSUE #7846: Bare Workflow - expo-updates crashes on iOS and error on Android.
# 👇 The answer to this issue 👇

Please read this comment: https://github.com/expo/expo/issues/7846#issuecomment-629720280

<hr />

# The original report

### Environment

  Expo CLI 3.17.24 environment info:
    System:
      OS: macOS 10.15.4
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 13.7.0 - ~/.nvm/versions/node/v13.7.0/bin/node
      Yarn: 1.22.4 - /usr/local/bin/yarn
      npm: 6.13.6 - ~/.nvm/versions/node/v13.7.0/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Android Studio: 3.6 AI-192.7142.36.36.6308749
      Xcode: 11.4/11E146 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^37.0.6 => 37.0.7 
      react: ~16.9.0 => 16.9.0 
      react-native: ~0.61.5 => 0.61.5 
    npmGlobalPackages:
      expo-cli: 3.17.24

### Steps to Reproduce

- `expo init`
- Pick "minimal" in Bare Workflow
-  `yarn install expo-updates expo-asset`
- Add in **App.js** in a `useEffect` hook : `const update = await Updates.checkForUpdateAsync();`
- `expo publish`
- run`yarn ios` or `yarn android`.

### Expected Behavior

The app should run normally without any issue.
Indeed my folder ios/PROJECT/Supporting/ contains`app.bundle, app.manifest, Expo.plist` filled up and `AppDelegate.h` and `AppDelegate.m `are updated.

### Actual Behavior

It crashes on iOS with "*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[OS_dispatch_data writeToFile:options:error:]: nil file argument'"

I get an error on Android: [Error: The updates module controller has not been properly initialized. If you're using a development client, you cannot check for updates. Otherwise, make sure you have called the native method UpdatesController.initialize().]

### Reproducible Demo

- `expo init`
- Pick "minimal" in Bare Workflow
-  `yarn install expo-updates expo-asset`
- Add in **App.js** in a `useEffect` hook : `const update = await Updates.checkForUpdateAsync();`
- `expo publish`
- run`yarn ios` or `yarn android`.

### Questions

Does it crash on iOS because it is in "development mode" ?
Is `releasing the app` the only way to test both apps?

### PR DESCRIPTION
**Pull request by esamelson**
# Why

Fixes https://github.com/expo/expo/issues/7846 and follow-up described in https://github.com/expo/expo/issues/7846#issuecomment-629720280

# How

- check for the correct condition so that we don't crash in development mode on iOS. Instead, we reject the promise with a note that the module can't be used in development mode.
- add a note/warning to the API page in docs about testing in release mode

# Test Plan

Repro from https://github.com/expo/expo/issues/7846#issuecomment-629720280 no longer crashes after this fix.

@brentvatne I'm adding you here mostly to get your opinion on if this experience (reject promise + docs warning) seems reasonable or if we should add more info elsewhere.

### CODE REVIEW

File: packages/expo-updates/src/Updates.ts [hunk 1]
```diff
@@ -32,13 +32,27 @@ export async function reloadAsync(): Promise<void> {
...
 
 export async function checkForUpdateAsync(): Promise<UpdateCheckResult> {
   if (!ExpoUpdates.checkForUpdateAsync) {
     throw new UnavailabilityError('Updates', 'checkForUpdateAsync');
   }
+  if (__DEV__) {
+    throw new CodedError(
+      'ERR_UPDATES_DISABLED',
+      'You cannot check for updates in development mode. To test manual updates, make a ' +
+        'release build with `yarn ios --configuration Release` or `yarn android --variant Release`.'
```
Comment (brentvatne) [inline packages/expo-updates/src/Updates.ts:53]: probably don't want to couple this to yarn, `npm` is a bit more agnostic

**Review by brentvatne — APPROVED** [review]

File: packages/expo-updates/src/Updates.ts (hunk 1 above)
Comment (objectiveSee) [reply packages/expo-updates/src/Updates.ts]: FWIW I don't use `yarn` so this build command doesn't help me. It's odd that this doesn't work in Development (and just crashes). I am seeing this on `expo-updates` ~0.1.2`. Will just test in production when I get a chance :)

Comment (brentvatne) [reply packages/expo-updates/src/Updates.ts]: if you see the changes after the update you will notice we changed it to `npm`

so, `npm run ios --configuration Release` and the equivalent for android

Comment (brentvatne) [reply packages/expo-updates/src/Updates.ts]: i would recommend upgrading to 0.2.4 of expo-updates, more info on it here https://blog.expo.io/over-the-air-updates-from-expo-are-now-even-easier-to-use-376e2213fabf :)

**Pull request merged**